Add inversion mixing and internal conduction

pmskintner opened the pull request:
To introduce more realism into HPWHsim mixing of temperature inversions and internal conduction between nodes is added. For details see the commit. 

Graphs below come are from the DHW_inv test in cse, showing the 6 thermocouples before and after the changes were made.

![image](https://user-images.githubusercontent.com/50156325/61082071-91c1ae00-a3dd-11e9-94f9-857deffaacc1.png)

![image](https://user-images.githubusercontent.com/50156325/61082146-b6b62100-a3dd-11e9-8212-a0d3b3903d51.png)

pmskintner committed:
Update HPWH.cc

pmskintner committed:
Update HPWH.hh

pmskintner committed:
Update HPWH.cc

pmskintner committed:
Add inversion mixing and internal conduction

To introduce more realism into the HPWHsim mixing of temperature inversions and internal conduction between nodes is added.

- Inversion mixing is done at the end of runOneStep after all of the other changes to the tank temperature. It starts at the top of the tank and moves down checking for an inversion and, if it finds one, averages the temperatures between all inverted nodes.

- Internal conduction happens at the end of updateTankTemps after the heaters have input heat. Conduction is calculated using a 1D finite difference method with explicit time stepping, where the boundary conditions use a fraction of the UA for the top and bottom surfaces.

- All hpwh tanks are assumed to be 1.2m tall and the radius is found using the tank volume.

- New features are logically controlled by doInversionMixing and doConduction they default to true, and for a hpwh can be change with setDoInversionMixing() and setDoConduction() respectively.

- With doConduction = false, the HPWH will use the old method of calculating UAlosses based on the average tank temperature

- New tests were created to test conduction and inversion mixing, testCondu and testRec.

- Added statements to test30 to make doConduction and doInversionMixing are false, and statements to test50 so that doConduction is false. When running the tests test30 will pass all regresssion tests and test50 will pass all regression tests except for the Sanden unit.

pmskintner committed:
New tests made

pmskintner committed:
Add Sanden Combi Test

Test to check against the Sanden experiment where a small inversion occurs.

pmskintner committed:
Updated to Version 1.5.1

pmskintner committed:
Fix of tempTemps memory issue

pmskintner committed:
Increment Version Number

pmskintner committed:
Check that UA is between 0 and 500.

Added If to check UA in checkInputs()

Incremented Version Number

pmskintner committed:
Check on upper limit of EF

pmskintner committed:
Update HPWH.cc

nealkruis committed:
Merge pull request #48 from EcotopeResearch/limit_UA

Limit UA

nealkruis commented:
@pmskintner looking at the charts, it looks like the inlet water in one case is 32 C, and the other is 25 C. Am I reading this correctly? Can you attach the CSE test file so I can see how things are changing?

pmskintner commented:
I didn't change anything in the tests, so the input should be 45F in both cases. Attached is a zip with the .cse test file and the output in .csv. Also below there's a plot where I added the inlet water temperature. The flat section around minute 800 is after the hot water is injected on bottom and the water heater is running the top resistance element. 

![image](https://user-images.githubusercontent.com/50156325/63125808-6c3b3d80-bf63-11e9-9089-8bd24d4e2a38.png)


[DHW_INV.zip](https://github.com/EcotopeResearch/HPWHsim/files/3507120/DHW_INV.zip)

pmskintner committed:
Updated from Master (includes versioning)

pmskintner committed:
Updated from Master (Removed Conflicts)

pmskintner committed:
Removed Conflicts in HPWH.in.hh

pmskintner committed:
Fix Warnings

pmskintner committed:
Move invalid UA checking in restank

pmskintner committed:
Update to tests and reference tests

nealkruis commented on src/HPWH.in.hh:
```diff
@@ -449,6 +458,8 @@ class HPWH {
 	/**< the setpoint of the tank  */
 	double *tankTemps_C;
 	/**< an array holding the temperature of each node - 0 is the bottom node, numNodes is the top  */
+	double *tempTemps;
```
Maybe come up with a better name? it is still in Celsius, so maybe "nextTankTemps_C"?

nealkruis commented on test/main.cc:
```diff
@@ -263,12 +274,26 @@ int main(int argc, char *argv[])
       drStatus = HPWH::DR_ENGAGE;
     }
 
+	//// Let's create a feedback so the inlet temperature is the outlet temperature - 5 minutes
```
Remove these lines?

nealkruis commented on test/main.cc:
```diff
@@ -263,12 +274,26 @@ int main(int argc, char *argv[])
                       allSchedules[3][i],  // External Temp (C)
-                      drStatus, 1.0);    // DDR Status (now an enum. Fixed for now as allow)
+                      drStatus, // DDR Status (now an enum. Fixed for now as allow)
+					  1.0);    // Minutes per step
```
What's going on with the whitespace here?

pmskintner replied:
Not sure exactly what you mean, I tried respacing it out if it was showing up poorly formatted. But if you hate seeing white I can remove it, I think it adds to readability.

nealkruis replied:
I was just talking about how the last line is indented so much further. Sometimes things show up this way in GitHub if there isn't consistent use of tabs vs. spaces.

nealkruis commented on src/HPWH.cc:
```diff
@@ -391,13 +403,14 @@ int HPWH::runOneStep(double inletT_C, double drawVolume_L,
 	}
 
+	// check for inverted temperature profile
+	if (doInversionMixing) 	mixTankInversions();
```
put on new line with scoping brackets.

nealkruis commented on src/HPWH.cc:
```diff
@@ -1207,22 +1227,108 @@ void HPWH::updateTankTemps(double drawVolume_L, double inletT_C, double tankAmbi
+	// on the top and bottom node being the fraction of UA that corresponds to the top and bottom of the tank.  
+	if (doConduction) {
+		// height estimate from Rheem along with the volume is used to get the radius and node_height
+		double height = 1.2; //meters
```
`static const double height = 1.2; // meters`

nealkruis commented on src/HPWH.cc:
```diff
@@ -1207,22 +1227,108 @@ void HPWH::updateTankTemps(double drawVolume_L, double inletT_C, double tankAmbi
+	if (doConduction) {
+		// height estimate from Rheem along with the volume is used to get the radius and node_height
+		double height = 1.2; //meters
+		double rad = sqrt(tankVolume_L / (1000 * 3.14159 * height));
```
`const double rad...`. There are a lot of things here you can probably make constants (i.e., they don't change within the scope).

nealkruis commented on src/HPWH.cc:
```diff
@@ -1207,22 +1227,108 @@ void HPWH::updateTankTemps(double drawVolume_L, double inletT_C, double tankAmbi
+
+		// Get the "constant" tau for the stability condition and the conduction calculation
+		double tau = KWATER_WpermC / ( CPWATER_kJperkgC * 1000 * DENSITYWATER_kgperL * 1000 * (node_height * node_height) ) * minutesPerStep * 60.0;
+		if (tau > 0.5) msg("The stability condition for conduction has failed, these results are going to be interesting!\n");
```
new line + bracket for "if" statement

nealkruis commented on src/HPWH.cc:
```diff
@@ -1207,22 +1227,108 @@ void HPWH::updateTankTemps(double drawVolume_L, double inletT_C, double tankAmbi
+	}
+	else { // Ignore tank conduction and calculate UA losses in the old way.
+		
+		for (int i = 0; i < numNodes; i++) 	tempTemps[i] = tankTemps_C[i];
```
new line + bracket for "if" statement (check throughout).

nealkruis commented on src/HPWH.cc:
```diff
@@ -1207,22 +1227,108 @@ void HPWH::updateTankTemps(double drawVolume_L, double inletT_C, double tankAmbi
+			tempTemps[i] -= standbyLosses_kJ / ((volPerNode_LperNode * DENSITYWATER_kgperL) * CPWATER_kJperkgC);
+		}
+	}
+	else { // Ignore tank conduction and calculate UA losses in the old way.
```
It's not clear we would want to ever use the old UA loss method. If we did, it probably shouldn't be compounded with how/if we are calculating conduction.

pmskintner replied:
So I changed it, but it's pretty screwy. Because the top node loses heat faster than the other nodes it constantly creates inversions. I made two plots below, the first is temperature, note the green line is the top node, and the second is the output power and the hot water draws. This also seems to result in weird behavior for the compressor.

I'm fine leaving this in as a setting for the research build but just thought you should be warned.  

![image](https://user-images.githubusercontent.com/50156325/63392790-fcff8800-c36c-11e9-846a-ea1f275d355a.png)

![image](https://user-images.githubusercontent.com/50156325/63392792-fffa7880-c36c-11e9-9826-6489051dc2c3.png)

nealkruis replied:
It seems like we should always do inversion mixing, too.

nealkruis commented on src/HPWH.cc:
```diff
@@ -1207,22 +1227,108 @@ void HPWH::updateTankTemps(double drawVolume_L, double inletT_C, double tankAmbi
+// Inversion mixing modeled after bigladder EnergyPlus code PK
+void HPWH::mixTankInversions() {
+	bool hasInversion;
+	double volumePerNode_L = tankVolume_L / numNodes;
```
`const`

nealkruis commented on src/HPWH.cc:
```diff
@@ -2682,9 +2805,14 @@ int HPWH::HPWHinit_resTank(double tankVol_L, double energyFactor, double upperPo
+		if (hpwhVerbosity >= VRB_reluctant && tankUA_kJperHrC < -0.1) {
+			msg("Computed tankUA_kJperHrC is less than 0, and is reset to 0.");
+		}
+		tankUA_kJperHrC = 0;
```
Generally, it's good practice to always add a decimal point to floats and doubles so you don't accidentally do something where the compiler mistakes it for an integer. E.g., `float x = 2/3` makes x equal 0.0 (should be `float x = 2./3.` instead).

nealkruis commented on src/HPWH.cc:
```diff
@@ -2729,6 +2857,8 @@ int HPWH::HPWHinit_genericHPWH(double tankVol_L, double energyFactor, double res
 	tankTemps_C = new double[numNodes];
 	setpoint_C = F_TO_C(127.0);
 
+	tempTemps = new double[numNodes];
```
We should look into converting all the double[] arrays to std::vector to avoid potential memory problems.

chipbarnaby commented on src/HPWH.cc:
```diff
@@ -2139,6 +2246,12 @@ int HPWH::checkInputs(){
+	//Check if the UA is out of bounds
+	if (tankUA_kJperHrC < 0.0) {
+		msg("The tankUA_kJperHrC must be above 0 for a HPWH, tankUA_kJperHrC is: %f  \n", tankUA_kJperHrC);
+		returnVal = HPWH_ABORT;
```
Imprecise message -- test is < 0, but message says "must be above 0".

chipbarnaby commented on src/HPWH.cc:
```diff
@@ -2151,6 +2264,7 @@ int HPWH::HPWHinit_file(string configFile){
 	if (tankTemps_C != NULL) delete[] tankTemps_C;
+	if (tempTemps != NULL) delete[] tempTemps;
 	if (setOfSources != NULL) delete[] setOfSources;
 
```
if (p != NULL) delete[] p -- the NULL test is not needed, delete[] does nothing when p == NULL.  However, depending on the situation, it is often desirable to add p = NULL after the delete[] to be absolutely sure p will not be delete[]ed again.  Or use stdlib smart pointers.

Also, there seem to be several places where these arrays are allocated or deleted.  Should there be functions such as AllocArrays() and DeleteArrays() that are called from the several places?  That approach makes changes easier/safer -- no need to review the entire file, just change the functions.

pmskintner replied:
There's many places these arrays are allocated and deleted, and I'm not exactly sure why for each case. I'll remove NULL tests, but going through the entire file can probably wait.

chipbarnaby commented on src/HPWH.in.hh:
```diff
@@ -475,6 +486,12 @@ class HPWH {
 
+  bool doInversionMixing;
+  /**<  whether or not to model will mix temperature inversions in the tank  */
+
```
Comments of this sort are ambiguous.  Something like the following is better.
// iff true, model inversion mixing <whatever>

where iff = if and only if

chipbarnaby reviewed (approved):
Generally looks good.  Several picky comments interleaved below.

I realize the code was inherited from others, warts and all.  We should all keep discussing preferred style and try to leave the code in better shape each time it gets worked on.


General style -- put if stuff on separate line
  if (xxx)
      doWhatever()

nealkruis reviewed (changes requested):
Looking pretty good. A little clean up here and there.

pmskintner committed:
Most of the Reviewer Changes

More to come

pmskintner committed:
Changed if's to bracketed form

pmskintner committed:
Replaced old UA loss method with new

Changed ref tests that were not using internal conduction and thus used the old UA loss method as well
